### PR TITLE
Append HTTP header correctly

### PR DESCRIPTION
**Pull request by mhoeger**
Resolves https://github.com/Azure/azure-functions-host/issues/4358

### CODE REVIEW

**Comment by mhoeger** [comment]
@lmolkova and @brettsam - there was an issue where users who set the "request-context" header had their function apps stop working due to an unsafe "Add" of user headers to the http response (#4358). 

Does this look like the right solution? Will this maintain App Insights functionality and user-added functionality?

**Comment by lmolkova** [comment]
@mhoeger agreed it should not throw. I'm curious why would user inject request-context into response themselves. This is AI internal header and we don't expect anyone to interfere with it.
As a workaround, I suggest that users should remove code that deals with request-context in their code.

Having two comma separated request-context headers does not sound like the right approach as well (better though because it does not throw), because AppInsights SDKs might not be able to work correctly with two headers. 

I suggest that another header should not be added if the key is already present.

BTW Request-Context is the only **response** header AppInsights injects.

**Comment by mhoeger** [comment]
@lmolkova - the customer in this case called one function from another function and kept appending to the response object. 

What would be the correct behavior in this case? To take the request-context that is injected by app insights and ignore the one that, from our perspective, could have been set by the customer? Or to take the customer-set one and ignore the injected request-context?

**Comment by lmolkova** [comment]
Did customer try to fix some issue with ApplicaitonInsights by injecting request-context header or implement some missing functionality? 
If it is the workaround for the issue - let us know what the issue is, so we can fix it. If this was a missing feature - now it is implemented out of the box in WebJobs, customer should stop injecting request-context header in their code. 

I'll let @brettsam or @fabiocav comment on what is the best way to proceed from webjobs perspective regardless of AppInsights specific scenario.

**Comment by brettsam** [comment]
@mhoeger can correct me if I'm wrong... but from what I understand, they're getting that header from the response back from calling another function app. 

Client -> FunctionApp1 -> FunctionApp2

But because this is a Node app, the Functions host takes the headers from the Node process and apply them to the .NET response, which is what is blowing up because App Insights has already injected that header.

How does App Insights handle the header injection if the request-context header is already there? We should probably do something similar...

**Comment by lmolkova** [comment]
@brettsam thanks for the explanation! Now I see what's going on here.

AppInsights just overrides the whole header:  
https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/4e05f2aa23b6eea6a23d06fe28cf5fdcd48dd43d/src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HttpHeadersUtilities.cs#L80

Another reasonable approach could be to skip adding headers that are already present.

Appending another value to an existing header may result in AppInsights features related to correlation across multiple AppInsights resources broken as AppInsights never expects multiple values there.

Going forward, we are working on removing the need for Request-Context header and stopping sending it in requests and responses.

**Comment by mhoeger** [comment]
Good to know, thanks!

@brettsam and @fabiocav - in the host, presumably if we inject response headers, they are important. Is that ok to override/skip customer-set headers if they conflict with injected response headers? Perhaps logging a warning if we override a header?

**Comment by brettsam** [comment]
@lmolkova -- when does this header get added by App Insights? Will it overwrite anything the customer adds or do they have a chance to override it themselves somehow?

My current thoughts: appending seems like the wrong approach. And since customers have the ability to disable the auto-injection by App Insights, I think that we should use the one that App Insights has injected and ignore any that came from the Language Worker. I don't know of any other headers that we auto-inject -- should we special-case this one?

**Comment by lmolkova** [comment]
@brettsam just wanted to clarify: customers should never add request-context themselves - it is not supported and the header content/format is not public or documented. 
Different AppInsights implementations in different languages might behave differently, but node and .NET one inject this header when request *starts*. So in theory customers might override it, but as I mentioned - they should never do this.

So, in all cases, this is AppInsights that injects the header. If it behaves correctly, language worker will inject EXACTLY the same header as host, so for the AppInsights case, it should not matter which one to use.

This scenario is very special: users enable AppInsights in node.js functions and track requests with it (in addition to the .NET host) because they want to have a correlation between node.js code and .net host telemtery. This causes another header injection. Once https://github.com/Azure/azure-functions-host/issues/3747 is solved, the recommendation would be to stop tracking requests in language workers.

So, from AppInsights perspective - it does not really matter which approach is used (as long as there is one value) and in the long term, there is no concern at all (header will be on the deprecation path soon).

**Comment by lmolkova** [comment]
>  I don't know of any other headers that we auto-inject -- should we special-case this one?

This is the only response header AppInsights injects. I'd prefer to not special-case it and just have the same treatment for all cases like that

**Comment by mhoeger** [comment]
@lmolkova - would appending an extra header (not a comma-separated header value) still cause problems? Something like: 
```
Date →Tue, 30 Apr 2019 18:47:06 GMT
Content-Type →text/plain; charset=utf-8
Server →Kestrel
Content-Length →11
Request-Context →appId=cid-v1:229f9314-81fc-4393-8fac-38c02fd9a41e
Request-Context →mycustomvalue
```

**Comment by lmolkova** [comment]
this is equivalent for appending comma-separated value from HTTP RFC perspective and causes the same issues.

**Comment by fabiocav** [comment]
Sorry... approved before checking the conversation here first.

I think the approach of ignoring headers that have already been added to the response is reasonable. I agree that taking an approach where we end up with multiple values can lead to unpredictable and undesirable results. 

I also want to get clarification becasue @brettsam's description of the flow doesn't match what @mhoeger had stated before. Is this a result of calling another function app and using the response headers, or is it a conflict with another framework adding that header completely outside of an AI context? Based on Marie's description, it was the later, but what has been described here makes sense.

**Comment by brettsam** [comment]
@mhoeger's initial description was what we thought was happening -- but it turns out it wasn't another library adding this header, but app insights itself (since the customer's app was calling into another Function app).

I also think we need to skip the header if it's already there. And probably log something so we know it's happening? 

Customers can still turn off this header injection if they don't want this behavior. But I'm guessing that in every case it will be App Insights adding it via a separate function app like we're seeing here.

**Comment by mhoeger** [comment]
@fabiocav - I tried implementing the middleware approach you talked about since it avoids any possibility of a breaking change. Let me know how this looks, or if it adds too much unnecessary complexity!

File: src/WebJobs.Script.WebHost/Middleware/ResponseContextItemsCheckMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            if (_next != null)
+            {
+                await _next(context);
+            }
+
+            // Check response for duplicate http headers
+            object value = null;
+            if (context?.Items?.TryGetValue(ScriptConstants.AzureFunctionsDuplicateHttpHeaderKey, out value) ?? false)
+            {
+                _logger.LogInformation($"Duplicate HTTP header from function invocation removed. Key: '{value.ToString() ?? string.Empty}'.");
```
Comment (brettsam) [inline src/WebJobs.Script.WebHost/Middleware/ResponseContextItemsCheckMiddleware.cs]: Should we log this as Debug? If someone has App Insights set up (with an out-of-proc language) and call a downstream function app, they'll see this log with every request. With Debug, we'll see it in our logs still.

**Comment by mhoeger** [comment]
@fabiocav - addressed changes! mind re-reviewing?

**Pull request merged**